File: Serialization/Go/stack/stack.go
package main

import (
	"encoding/binary"
	"errors"
	"fmt"
	"io"
	"bufio"
	"os"
	"strconv"
)

const MAX_SIZE = 10

type SNode struct {
	key  string
	next *SNode
}

type Stack struct {
	head *SNode
	size int
}

func NewStack() *Stack {
	return &Stack{
	 head: nil,
	 size: 0,
	}
}

func NewStackFromSlice(items ...string) *Stack {
	s := NewStack()
	for _, item := range items {
		s.Push(item)
	}
	return s
}

func (s *Stack) Push(data string) error {
	if s.size >= MAX_SIZE {
	 return errors.New("stack overflow: maximum size reached")
	}
	
	newNode := &SNode{
	 key:  data,
	 next: s.head,
	}
	s.head = newNode
	s.size++
	return nil
}

func (s *Stack) Pop() (string, error) {
	if s.head == nil {
	 return "", errors.New("stack underflow: stack is empty")
	}
	
	data := s.head.key
	s.head = s.head.next
	s.size--
	return data, nil
}

func (s *Stack) IsEmpty() bool {
	return s.head == nil
}

func (s *Stack) GetSize() int {
	return s.size
}

func (s *Stack) WriteBinary(filename string) error {
	file, err := os.Create(filename)
	if err != nil {
	 return fmt.Errorf("не удалось открыть файл для записи: %s", filename)
	}
	defer file.Close()

	if err := binary.Write(file, binary.LittleEndian, int32(s.size)); err != nil {
	 return err
	}

	current := s.head
	for current != nil {
	 keyBytes := []byte(current.key)
	 if err := binary.Write(file, binary.LittleEndian, int32(len(keyBytes))); err != nil {
	  return err
	 }
	 if _, err := file.Write(keyBytes); err != nil {
	  return err
	 }
	 current = current.next
	}
	
	return nil
}

func (s *Stack) ReadBinary(filename string) error {
	file, err := os.Open(filename)
	if err != nil {
	 return fmt.Errorf("не удалось открыть файл: %s", filename)
	}
	defer file.Close()

	s.Clear()

	var fileSize int32
	if err := binary.Read(file, binary.LittleEndian, &fileSize); err != nil {
	 return err
	}

	if fileSize > MAX_SIZE {
	 return errors.New("размер стека в файле превышает максимально допустимый")
	}

	tempArray := make([]string, fileSize)
	
	for i := int(fileSize) - 1; i >= 0; i-- {
	 var keyLength int32
	 if err := binary.Read(file, binary.LittleEndian, &keyLength); err != nil {
	  return errors.New("ошибка чтения длины строки из файла")
	 }
	 
	 keyBytes := make([]byte, keyLength)
	 if _, err := io.ReadFull(file, keyBytes); err != nil {
	  return errors.New("ошибка чтения строки из файла")
	 }
	 tempArray[i] = string(keyBytes)
	}

	for i := 0; i < int(fileSize); i++ {
	 if err := s.Push(tempArray[i]); err != nil {
	  return err
	 }
	}

	return nil
}

func (s *Stack) WriteText(filename string) error {
	file, err := os.Create(filename)
	if err != nil {
	 return fmt.Errorf("не удалось открыть файл для записи: %s", filename)
	}
	defer file.Close()

	if _, err := file.WriteString(fmt.Sprintf("%d\n", s.size)); err != nil {
	 return err
	}

	current := s.head
	for current != nil {
	 if _, err := file.WriteString(current.key + "\n"); err != nil {
	  return err
	 }
	 current = current.next
	}
	
	return nil
}

func (s *Stack) ReadText(filename string) error {
	file, err := os.Open(filename)
	if err != nil {
		return fmt.Errorf("не удалось открыть файл: %s", filename)
	}
	defer file.Close()

	s.Clear()

	scanner := bufio.NewScanner(file)
	
	if !scanner.Scan() {
		return errors.New("не удалось прочитать размер стека")
	}
	
	sizeStr := scanner.Text()
	fileSize, err := strconv.Atoi(sizeStr)
	if err != nil {
		return err
	}

	if fileSize > MAX_SIZE {
		return errors.New("размер стека в файле превышает максимально допустимый")
	}

	tempArray := make([]string, fileSize)
	
	for i := fileSize - 1; i >= 0; i-- {
		if !scanner.Scan() {
			return errors.New("не удалось прочитать элемент стека")
		}
		tempArray[i] = scanner.Text()
	}

	for i := 0; i < fileSize; i++ {
		if err := s.Push(tempArray[i]); err != nil {
			return err
		}
	}

	return scanner.Err()
}

func (s *Stack) Print() {
	if s.IsEmpty() {
	 fmt.Println("Стек пуст")
	 return
	}
	
	current := s.head
	for current != nil {
	 fmt.Print(current.key, " ")
	 current = current.next
	}
	fmt.Println()
}

func (s *Stack) Clear() {
	for !s.IsEmpty() {
	 s.Pop()
	}
}

func main() {	
	a := NewStackFromSlice("HELP", "I CANT", "HOLD IT", "ANYMORE")
	a.Print()
	a.WriteBinary("test")
	a.WriteText("test2")

	b := NewStack()
	b.ReadBinary("test")
	b.Print()

	c := NewStack()
	c.ReadText("test2")
	c.Print()
}

File: Serialization/C++/Stack/Stack.cpp
#include "Stack.h"

SNode* Stack::createSNode(string key){
    SNode* newSNode = new SNode;
    newSNode->key = key;
    newSNode->next = nullptr;
    return newSNode;
}

Stack::Stack() : head(nullptr)
               , size(0) {}

Stack::Stack(initializer_list<string> list) : head(nullptr)
                                            , size(0) {
    for (string item : list) {
        push(item);
    }
}

void Stack::push(string data){
    if (this->size >= this->MAX_SIZE){
        throw overflow_error("Stack overflow: Maximum size reached");
    }
    SNode* newSNode = createSNode(data);
    newSNode->next = this->head;
    this->head = newSNode;
    this->size++;
}

string Stack::pop(){
    if (this->head == nullptr){
        throw underflow_error("Stack underflow: Stack is empty");
    }
    
    SNode* oldhead = this->head;
    string data = oldhead->key;
    this->head = this->head->next;
    delete oldhead;
    this->size--;
    return data;
}

bool Stack::isEmpty() {
    return this->head == nullptr;
}

int Stack::getSize() {
    return this->size;
}

void Stack::writeBinary(const string& filename) {
    ofstream file(filename, ios::binary);
    if (!file) {
        throw runtime_error("Не удалось открыть файл для записи: " + filename);
    }

    file.write(reinterpret_cast<const char*>(&this->size), sizeof(this->size));

    SNode* current = head;
    while (current != nullptr) {
        size_t keyLength = current->key.length();
        file.write(reinterpret_cast<const char*>(&keyLength), sizeof(keyLength));
        file.write(current->key.c_str(), keyLength);
        current = current->next;
    }
    
    file.close();
}

void Stack::readBinary(const string& filename) {
    ifstream file(filename, ios::binary);
    if (!file) {
        throw runtime_error("Не удалось открыть файл: " + filename);
    }

    while (!isEmpty()) {
        pop();
    }

    int fileSize;
    file.read(reinterpret_cast<char*>(&fileSize), sizeof(fileSize));

    string* tempArray = new string[fileSize];
    
    for (int i = fileSize - 1; i >= 0; --i) {
        size_t keyLength;
        if (!file.read(reinterpret_cast<char*>(&keyLength), sizeof(keyLength))) {
            delete[] tempArray;
            throw runtime_error("Ошибка чтения длины строки из файла");
        }
        
        tempArray[i].resize(keyLength);
        if (!file.read(&tempArray[i][0], keyLength)) {
            delete[] tempArray;
            throw runtime_error("Ошибка чтения строки из файла");
        }
    }

    for (int i = 0; i < fileSize; ++i) {
        push(tempArray[i]);
    }

    delete[] tempArray;
    file.close();
}

void Stack::writeText(const string& filename) {
    ofstream file(filename);
    if (!file) {
        throw runtime_error("Не удалось открыть файл для записи: " + filename);
    }

    file << this->size << endl;

    SNode* current = head;
    while (current != nullptr) {
        file << current->key << endl;
        current = current->next;
    }
    
    file.close();
}

void Stack::readText(const string& filename) {
    ifstream file(filename);
    if (!file) {
        throw runtime_error("Не удалось открыть файл: " + filename);
    }

    while (!isEmpty()) {
        pop();
    }

    string sizeStr;
    getline(file, sizeStr);
    int fileSize = stoi(sizeStr);

    if (fileSize > MAX_SIZE) {
        throw runtime_error("Размер стека в файле превышает максимально допустимый");
    }

    string* tempArray = new string[fileSize];
    
    for (int i = fileSize - 1; i >= 0; --i) {
        getline(file, tempArray[i]);
    }

    for (int i = 0; i < fileSize; ++i) {
        push(tempArray[i]);
    }

    delete[] tempArray;
    file.close();
}

void Stack::print() {
    if (isEmpty()) {
        cout << "Стек пуст" << endl;
        return;
    }
    
    SNode* current = head;
    while (current != nullptr) {
        cout << current->key << " ";
        current = current->next;
    }
    cout << endl;
}